lib/ex_unit/test/ex_unit/capture_io_test.exs

```diff
@@ -86,6 +86,102 @@ defmodule ExUnit.CaptureIOTest do
            end) == "a"
   end
 
+  test "async capture_io works with put chars to stderr" do
+    me = self()
+
+    Enum.each(1..6, fn num ->
+      spawn(fn ->
+        captured =
+          capture_io(:stderr, fn ->
+            :io.put_chars(:standard_error, "#{num}\n")
+            Process.sleep(10)
+            :io.put_chars(:standard_error, "#{num + 10}\n")
+          end)
+
+        send(me, captured)
+      end)
+
+      Process.sleep(1)
+    end)
+
+    expected = [
+      {~r/^1\n/, ~r/11\n/},
+      {~r/^2\n/, ~r/12\n/},
+      {~r/^3\n/, ~r/13\n/},
+      {~r/^4\n/, ~r/14\n/},
+      {~r/^5\n/, ~r/15\n/},
+      {~r/^6\n/, ~r/16\n/}
+    ]
+
+    assert_received_regex(expected)
+  end
+
+  test "raises when async capturing a named device with a different encoding than the first" do
+    me = self()
+
+    Enum.each([:latin1, :unicode, :latin1], fn encoding ->
+      spawn(fn ->
+        try do
+          capture_io(:stderr, [encoding: encoding], fn ->
+            :io.put_chars(:standard_error, "a")
+            Process.sleep(3)
+          end)
+        rescue
+          e in [ArgumentError] ->
+            send(me, e.message)
+        end
+      end)
+    end)
+
+    assert_receive "attempted to change the encoding for a currently captured device :standard_error.\n\nCurrently set as: :latin1\nGiven: :unicode" <>
+                     _
+
+    refute_receive "attempted to change the encoding for a currently captured device :standard_error.\n\nCurrently set as: :unicode\nGiven: :latin1" <>
+                     _
+  end
+
+  test "raises when async capturing a named device with an input given to an already captured device" do
+    me = self()
+
+    Enum.each(["first", "second", ""], fn input ->
+      spawn(fn ->
+        try do
+          capture_io(:stderr, [input: input], fn ->
+            :io.put_chars(:standard_error, "a")
+            Process.sleep(5)
+          end)
+        rescue
+          e in [ArgumentError] ->
+            send(me, e.message)
+        end
+      end)
+    end)
+
+    assert_receive "attempted to give an input \"second\" for a currently captured device :standard_error. If you need to give an input to a captured device, you cannot run your test asynchronously"
+
+    refute_receive "attempted to give an input \"\" for a currently captured device :standard_error" <>
```

Comment:
Negative assertions are expensive (they always wait) and brittle (if someone change the error message and forget to update it, then it will always be wrong. Instead, do a `send(me, {:ok, input})` after `capture_io` and `assert_receive` on that instead.